### ISSUE TITLE
"Training on the GPU": "RuntimeError: Need input of dimension 4..."

### ISSUE THREAD

**Issue by drscotthawley** [pytorch/tutorials#22]
Towards the end of "Deep Learning with PyTorch", there's the statement:

> Exercise: Try increasing the width of your network (argument 1 and 2 of nn.Conv2d, see what kind of speedup you get.

Following these instructions, going back and making such a change, and then clicking through the rest of the notebook, it eventually fails when you get to "3. Train the network", producing the errors such as those listed below.

The tutorial is unclear.  There are two nn.Conv2d layers given (self.conv1 and self.conv2), and the tutorial is unclear about which one should be changed.  So far, my changing *any* of the numbers for the first two parameters of these layers seems to produce errors.  For example, doubling the first two values of the second Conv2d line from (6,16,5) to (12,32,5) as in...
```
def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(3, 6, 5)
        self.pool  = nn.MaxPool2d(2,2)
        self.conv2 = nn.Conv2d(12, 32, 5)
        self.fc1   = nn.Linear(16*5*5, 120)
        self.fc2   = nn.Linear(120, 84)
        self.fc3   = nn.Linear(84, 10)
```

...produces these errors...

> ---------------------------------------------------------------------------
> RuntimeError                              Traceback (most recent call last)
> <ipython-input-60-b8de45dbc9f6> in <module>()
>      13 
>      14         # forward + backward + optimize
> ---> 15         outputs = net(inputs)
>      16         loss = criterion(outputs, labels)
>      17         loss.backward()
> 
> /usr/local/lib/python2.7/dist-packages/torch/nn/modules/module.pyc in __call__(self, *input, **kwargs)
>     208 
>     209     def __call__(self, *input, **kwargs):
> --> 210         result = self.forward(*input, **kwargs)
>     211         for hook in self._forward_hooks.values():
>     212             hook_result = hook(self, input, result)
> 
> <ipython-input-58-fed8675a8a25> in forward(self, x)
>      11     def forward(self, x):
>      12         x = self.pool(F.relu(self.conv1(x)))
> ---> 13         x = self.pool(F.relu(self.conv2(x)))
>      14         x = x.view(-1, 16*5*5)
>      15         x = F.relu(self.fc1(x))
> 
> /usr/local/lib/python2.7/dist-packages/torch/nn/modules/module.pyc in __call__(self, *input, **kwargs)
>     208 
>     209     def __call__(self, *input, **kwargs):
> --> 210         result = self.forward(*input, **kwargs)
>     211         for hook in self._forward_hooks.values():
>     212             hook_result = hook(self, input, result)
> 
> /usr/local/lib/python2.7/dist-packages/torch/nn/modules/conv.pyc in forward(self, input)
>     235     def forward(self, input):
>     236         return F.conv2d(input, self.weight, self.bias, self.stride,
> --> 237                         self.padding, self.dilation, self.groups)
>     238 
>     239 
> 
> /usr/local/lib/python2.7/dist-packages/torch/nn/functional.pyc in conv2d(input, weight, bias, stride, padding, dilation, groups)
>      35     f = ConvNd(_pair(stride), _pair(padding), _pair(dilation), False,
>      36                _pair(0), groups)
> ---> 37     return f(input, weight, bias) if bias is not None else f(input, weight)
>      38 
>      39 
> 
> /usr/local/lib/python2.7/dist-packages/torch/nn/_functions/conv.pyc in forward(self, input, weight, bias)
>      32         if k == 3:
>      33             input, weight = _view4d(input, weight)
> ---> 34         output = self._update_output(input, weight, bias)
>      35         if k == 3:
>      36             output, = _view3d(output)
> 
> /usr/local/lib/python2.7/dist-packages/torch/nn/_functions/conv.pyc in _update_output(self, input, weight, bias)
>      89 
>      90         self._bufs = [[] for g in range(self.groups)]
> ---> 91         return self._thnn('update_output', input, weight, bias)
>      92 
>      93     def _grad_input(self, input, weight, grad_output):
> 
> /usr/local/lib/python2.7/dist-packages/torch/nn/_functions/conv.pyc in _thnn(self, fn_name, input, weight, *args)
>     148         impl = _thnn_convs[self.thnn_class_name(input)]
>     149         if self.groups == 1:
> --> 150             return impl[fn_name](self, self._bufs[0], input, weight, *args)
>     151         else:
>     152             res = []
> 
> /usr/local/lib/python2.7/dist-packages/torch/nn/_functions/conv.pyc in call_update_output(self, bufs, input, weight, bias)
>     226         args = parse_arguments(self, fn.arguments[5:], bufs, kernel_size)
>     227         getattr(backend, fn.name)(backend.library_state, input, output, weight,
> --> 228                                   bias, *args)
>     229         return output
>     230     return call_update_output
> 
> RuntimeError: Need input of dimension 4 and input.size[1] == 12 but got input to be of shape: [4 x 6 x 14 x 14] at /home/soumith/local/builder/wheel/pytorch-src/torch/lib/THNN/generic/SpatialConvolutionMM.c:47
> 

Whereas doubling the first two parameters of the first nn.Conv2d line...

```
def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(6, 12, 5)
        self.pool  = nn.MaxPool2d(2,2)
        self.conv2 = nn.Conv2d(6, 16, 5)
        self.fc1   = nn.Linear(16*5*5, 120)
        self.fc2   = nn.Linear(120, 84)
        self.fc3   = nn.Linear(84, 10)

```

produces a similar error...  

> ---------------------------------------------------------------------------
> RuntimeError                              Traceback (most recent call last)
> <ipython-input-66-b8de45dbc9f6> in <module>()
>      13 
>      14         # forward + backward + optimize
> ---> 15         outputs = net(inputs)
>      16         loss = criterion(outputs, labels)
>      17         loss.backward()
> 
> /usr/local/lib/python2.7/dist-packages/torch/nn/modules/module.pyc in __call__(self, *input, **kwargs)
>     208 
>     209     def __call__(self, *input, **kwargs):
> --> 210         result = self.forward(*input, **kwargs)
>     211         for hook in self._forward_hooks.values():
>     212             hook_result = hook(self, input, result)
> 
> <ipython-input-64-74803b2fd4ea> in forward(self, x)
>      10 
>      11     def forward(self, x):
> ---> 12         x = self.pool(F.relu(self.conv1(x)))
>      13         x = self.pool(F.relu(self.conv2(x)))
>      14         x = x.view(-1, 16*5*5)
> 
> /usr/local/lib/python2.7/dist-packages/torch/nn/modules/module.pyc in __call__(self, *input, **kwargs)
>     208 
>     209     def __call__(self, *input, **kwargs):
> --> 210         result = self.forward(*input, **kwargs)
>     211         for hook in self._forward_hooks.values():
>     212             hook_result = hook(self, input, result)
> 
> /usr/local/lib/python2.7/dist-packages/torch/nn/modules/conv.pyc in forward(self, input)
>     235     def forward(self, input):
>     236         return F.conv2d(input, self.weight, self.bias, self.stride,
> --> 237                         self.padding, self.dilation, self.groups)
>     238 
>     239 
> 
> /usr/local/lib/python2.7/dist-packages/torch/nn/functional.pyc in conv2d(input, weight, bias, stride, padding, dilation, groups)
>      35     f = ConvNd(_pair(stride), _pair(padding), _pair(dilation), False,
>      36                _pair(0), groups)
> ---> 37     return f(input, weight, bias) if bias is not None else f(input, weight)
>      38 
>      39 
> 
> /usr/local/lib/python2.7/dist-packages/torch/nn/_functions/conv.pyc in forward(self, input, weight, bias)
>      32         if k == 3:
>      33             input, weight = _view4d(input, weight)
> ---> 34         output = self._update_output(input, weight, bias)
>      35         if k == 3:
>      36             output, = _view3d(output)
> 
> /usr/local/lib/python2.7/dist-packages/torch/nn/_functions/conv.pyc in _update_output(self, input, weight, bias)
>      89 
>      90         self._bufs = [[] for g in range(self.groups)]
> ---> 91         return self._thnn('update_output', input, weight, bias)
>      92 
>      93     def _grad_input(self, input, weight, grad_output):
> 
> /usr/local/lib/python2.7/dist-packages/torch/nn/_functions/conv.pyc in _thnn(self, fn_name, input, weight, *args)
>     148         impl = _thnn_convs[self.thnn_class_name(input)]
>     149         if self.groups == 1:
> --> 150             return impl[fn_name](self, self._bufs[0], input, weight, *args)
>     151         else:
>     152             res = []
> 
> /usr/local/lib/python2.7/dist-packages/torch/nn/_functions/conv.pyc in call_update_output(self, bufs, input, weight, bias)
>     226         args = parse_arguments(self, fn.arguments[5:], bufs, kernel_size)
>     227         getattr(backend, fn.name)(backend.library_state, input, output, weight,
> --> 228                                   bias, *args)
>     229         return output
>     230     return call_update_output
> 
> RuntimeError: Need input of dimension 4 and input.size[1] == 6 but got input to be of shape: [4 x 3 x 32 x 32] at /home/soumith/local/builder/wheel/pytorch-src/torch/lib/THNN/generic/SpatialConvolutionMM.c:47
> 
> 

Would love to know how to make this work.  Submitting as an Issue for the tutorial because following instructions produces errors.  
I'd suggest clarifying the instructions so that new adopters don't get stuck trying to resolve this.

**Comment by fmassa** [comment]
Hi,
Thanks for the feedback.
The [two first inputs to the convolution module](http://pytorch.org/docs/nn.html#torch.nn.Conv2d) correspond to the number of channels of the input image and the number of channel that will be returned.
In your case, you changed the number of output channels for one module, but you didn't change the number of input channels for the next convolutional module, which is still expecting to receive 6 channels.
In the same spirit, because you are changing the number of channels of the output of the convolution, the linear layer that comes afterwards will need to be adapted to the new number of channels as well.

**Comment by drscotthawley** [comment]
Thanks.  I figured, for an experienced user, it would be child's play to see the error and to fix it, but as a relative newbie just plodding along...     In Keras, one can in many cases change parameters of one layer and the system will automatically construct the "linkages" between the layers (although doing what I did would may also fail in Keras).

Perhaps the tutorial could more explicitly say something like...

"Exercise: Try increasing the width of your network (argument 2 of the first nn.Conv2d, and argument 1 of the second nn.Conv2d -- they need to be the same number), see what kind of speedup you get."

I'll fork, edit, and submit a pull request.

Just to be explicit, this sort of thing (changing both  6's to 12's) DOES work...
`    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(3, 12, 5)
        self.pool  = nn.MaxPool2d(2,2)
        self.conv2 = nn.Conv2d(12, 16, 5)
        self.fc1   = nn.Linear(16*5*5, 120)
        self.fc2   = nn.Linear(120, 84)
        self.fc3   = nn.Linear(84, 10)
`
